### PR TITLE
Anilist Module

### PR DESCRIPTION
**Pull request by Phyco-Ninja**
Created Anilit Module with use of anilist.co API. 
Capable of:

- Fetching Anime details using Anilist ID, MyAnimeList ID or even just by passing Anime Name
- Fetch Airing detail of Anime using Anilist ID or Anime Name. 
- Get all Scheduled animes
- Reverse Search Anime using tracemoepy
- Set your own Custom Templates for .anime
- Can provide Anime Character details tooo

DeepSource and PeP8 may go wild coz i have not tested this code yet.

### CODE REVIEW

File: plugins/anilist.py [hunk 1]
```diff
@@ -0,0 +1,630 @@
...
+    result = await return_json_senpai(anime_query, vars)
+    error = result.get('errors')
+    if error:
+        await CLOG.log(f"**ANILIST RETURNED FOLLOWING ERROR:**\n\n{error}")
+        error_sts = error[0].get('message')
+        await message.edit(f"**Error:** `[{error_sts}]`")
+        return
+
+    data = result.get('data')
+    data = data.get('Media')
```
Comment (rking32) [inline plugins/anilist.py:260]:
```suggestion
    data = result['data']['Media']
```
one line method

File: plugins/anilist.py [hunk 2]
```diff
@@ -0,0 +1,630 @@
...
+    duration = data.get('duration')
+    country = data.get('countryOfOrigin')
+    c_flag = flag.flag(country)
+    source = data.get('source')
+    coverImg = data.get('coverImage')['extraLarge']
+    bannerImg = data.get('bannerImage')
+    genres = data.get('genres')
+    genre = genres[0]
+    if len(genres) != 1:
+        genre = genres[0] + ", " + ", ".join(genres[1:])
```
Comment (rking32) [inline plugins/anilist.py:282]:
```suggestion
        genre = ", ".join(genres)
```
simply, this should work

File: plugins/anilist.py [hunk 3]
```diff
@@ -0,0 +1,630 @@
...
+        html_ += f"<b>Character ID</b>: {character['id']}<br>"
+        html_ += f"<h4>About Character and Role:</h4>{character.get('description', 'N/A')}"
+        html_char += f"{html_}<br><br>"
+
+    studios = ""
+    for studio in data['studios']['nodes']:
+        studios += "<a href='{}'>• {}</a> ".format(studio['siteUrl'], studio['name'])
+    url = data.get('siteUrl')
+
+    title_img = coverImg if coverImg else bannerImg
```
Comment (rking32) [inline plugins/anilist.py:312]:
```suggestion
    title_img = coverImg or bannerImg
```

File: plugins/anilist.py [hunk 4]
```diff
@@ -0,0 +1,630 @@
...
+    result = await return_json_senpai(anime_query, vars)
+    error = result.get('errors')
+    if error:
+        await CLOG.log(f"**ANILIST RETURNED FOLLOWING ERROR:**\n\n{error}")
+        error_sts = error[0].get('message')
+        await message.edit(f"**Error:** `[{error_sts}]`")
+        return
+
+    data = result.get('data')
+    data = data.get('Media')
```
Comment (rking32) [inline plugins/anilist.py:377]:
```suggestion
    data = result['data']['Media']
```

File: plugins/anilist.py [hunk 5]
```diff
@@ -0,0 +1,630 @@
...
+            'id': int(query),
+            'asHtml': True,
+            'type': "ANIME"
+        }
+    result = await return_json_senpai(anime_query, vars)
+    error = result.get('errors')
+    if error:
+        await CLOG.log(f"**ANILIST RETURNED FOLLOWING ERROR:**\n\n{error}")
+        error_sts = error[0].get('message')
+        await message.edit(f"**Error:** `[{error_sts}]`")
```
Comment (rking32) [inline plugins/anilist.py:373]: what about `message.err`

File: plugins/anilist.py [hunk 6]
```diff
@@ -0,0 +1,630 @@
...
+    status = data.get('status')
+    episodes = data.get('episodes')
+    country = data.get('countryOfOrigin')
+    c_flag = flag.flag(country)
+    source = data.get('source')
+    coverImg = data.get('coverImage')['extraLarge']
+    genres = data.get('genres')
+    genre = genres[0]
+    if len(genres) != 1:
+        genre = genres[0] + ", " + ", ".join(genres[1:])
```
Comment (rking32) [inline plugins/anilist.py:393]:
```suggestion
        genre = ", ".join(genres)
```

File: plugins/anilist.py [hunk 7]
```diff
@@ -0,0 +1,630 @@
...
+    out = f"[{c_flag}] {native} \n   ({title_})"
+    out += f"\n\n**ID:** `{id}`"
+    out += f"\n**Status:** `{status}`\n"
+    out += f"**Source:** `{source}`\n"
+    out += f"**Score:** `{score}`\n"
+    out += f"**Genres:** `{genre}`\n"
+    if air_on:
+        out += f"**Airing Episode:** `[{episode}/{episodes}]`\n"
+        out += f"\n`{air_on}`"
+    await message.reply_photo(coverImg, caption=out)
```
Comment (rking32) [inline plugins/anilist.py:411]: caption length should below `1024`
we should check it ?

File: plugins/anilist.py [hunk 8]
```diff
@@ -0,0 +1,630 @@
...
+    result = await return_json_senpai(airing_query, var)
+    error = result.get('errors')
+    if error:
+        await CLOG.log(f"**ANILIST RETURNED FOLLOWING ERROR:**\n\n{error}")
+        error_sts = error[0].get('message')
+        await message.edit(f"**Error:** `[{error_sts}]`")
+        return
+
+    data = result.get('data')
+    data = data['Page']['airingSchedules']
```
Comment (rking32) [inline plugins/anilist.py:438]:
```suggestion
    data = result['data']['Page']['airingSchedules']
```

File: plugins/anilist.py [hunk 9]
```diff
@@ -0,0 +1,630 @@
...
+    if len(genres) != 1:
+        genre = genres[0] + ", " + ", ".join(genres[1:])
+    score = data.get('averageScore')
+    air_on = None
+    if data['nextAiringEpisode']:
+        nextAir = data['nextAiringEpisode']['airingAt']
+        episode = data['nextAiringEpisode']['episode']
+        air_on = make_it_rw(nextAir, True)
+
+    title_ = english if english else romaji
```
Comment (rking32) [inline plugins/anilist.py:401]:
```suggestion
    title_ = english or romaji
```

File: plugins/anilist.py [hunk 10]
```diff
@@ -0,0 +1,630 @@
...
+    edited_ = False
+    out = ""
+    for air in data:
+        romaji = air['media']['title']['romaji']
+        english = air['media']['title']['english']
+        mid = air['mediaId']
+        epi_air = air['episode']
+        air_at = make_it_rw(air['airingAt'], True)
+        site = air['media']['siteUrl']
+        title_ = english if english else romaji
```
Comment (rking32) [inline plugins/anilist.py:450]:
```suggestion
        title_ = english or romaji
```

File: plugins/anilist.py [hunk 11]
```diff
@@ -0,0 +1,630 @@
...
+    'description': "Set your own Custom Anime Template "
+                   "that will be used to format .anime "
+                   "searches [<b>NOTE:</b> Requires "
+                   "proper key to be entered in curly braces]",
+    'usage': "{tr}setemp [Reply to text Message | Content]"})
+async def ani_save_template(message: Message):
+    text = message.input_str
+    replied = message.reply_to_message
+    if replied:
+        text = replied.text
```
Comment (rking32) [inline plugins/anilist.py:605]:
```suggestion
    text = message.input_or_reply_str
```

File: plugins/anilist.py [hunk 12]
```diff
@@ -0,0 +1,630 @@
...
+    template = await SAVED.find_one({'_id': "ANIME_TEMPLATE"})
+    if not template:
+        await SAVED.insert_one({'_id': "ANIME_TEMPLATE", 'anime_data': text})
+    else:
+        await SAVED.update_one({'_id': "ANIME_TEMPLATE"}, {"$set": {'anime_data': text}})
```
Comment (rking32) [inline plugins/anilist.py:610]:
```suggestion
    await SAVED.update_one({'_id': "ANIME_TEMPLATE"}, {"$set": {'anime_data': text}}, upsert=True)
```

**Comment by prono69** [comment]
OP

File: plugins/anilist.py [hunk 13]
```diff
@@ -0,0 +1,638 @@
...
+                'asHtml': True,
+                'type': "ANIME"
+            }
+
+    result = await return_json_senpai(anime_query, vars_)
+    error = result.get('errors')
+    if error:
+        await CLOG.log(f"**ANILIST RETURNED FOLLOWING ERROR:**\n\n`{error}`")
+        error_sts = error[0].get('message')
+        await message.err(f"`[{error_sts}]`")
```
Comment (rking32) [inline plugins/anilist.py:258]:
```suggestion
        await message.err(f"[{error_sts}]")
```

File: plugins/anilist.py [hunk 14]
```diff
@@ -0,0 +1,638 @@
...
+    html_pc += f"<a href='https://myanimelist.net/anime/{idmal}'>View on MAL</a>"
+    html_pc += f"<a href='{url}'> View on anilist.co</a>"
+    html_pc += f"<img src='{bannerImg}'/>"
+
+    title_h = english or romaji
+    synopsis_link = post_to_tp(title_h, html_pc)
+    try:
+        finals_ = ANIME_TEMPLATE.foramt(**locals())
+    except KeyError as kys:
+        await message.err(f"{kys}")
```
Comment (rking32) [inline plugins/anilist.py:337]:
```suggestion
        await message.err(kys)
```

File: plugins/anilist.py [hunk 15]
```diff
@@ -0,0 +1,638 @@
...
+            'id': int(query),
+            'asHtml': True,
+            'type': "ANIME"
+        }
+    result = await return_json_senpai(anime_query, vars_)
+    error = result.get('errors')
+    if error:
+        await CLOG.log(f"**ANILIST RETURNED FOLLOWING ERROR:**\n\n`{error}`")
+        error_sts = error[0].get('message')
+        await message.err(f"`[{error_sts}]`")
```
Comment (rking32) [inline plugins/anilist.py:375]:
```suggestion
        await message.err(f"[{error_sts}]")
```

File: plugins/anilist.py [hunk 16]
```diff
@@ -0,0 +1,638 @@
...
+    var = {'notYetAired': True}
+    await message.edit("`Fetching Scheduled Animes`\n\n"
+                       "[**NOTE:** __If Scheduled Animes exceeds limit. "
+                       "They will be forwarded to Log Channel to avoid spam.__]")
+    result = await return_json_senpai(airing_query, var)
+    error = result.get('errors')
+    if error:
+        await CLOG.log(f"**ANILIST RETURNED FOLLOWING ERROR:**\n\n{error}")
+        error_sts = error[0].get('message')
+        await message.err(f"`[{error_sts}]`")
```
Comment (rking32) [inline plugins/anilist.py:438]:
```suggestion
        await message.err(f"[{error_sts}]")
```

File: plugins/anilist.py [hunk 17]
```diff
@@ -0,0 +1,638 @@
...
+    var = {
+        'search': query,
+        'asHtml': True
+    }
+    result = await return_json_senpai(character_query, var)
+    error = result.get('errors')
+    if error:
+        await CLOG.log(f"**ANILIST RETURNED FOLLOWING ERROR:**\n\n`{error}`")
+        error_sts = error[0].get('message')
+        await message.err(f"`[{error_sts}]`")
```
Comment (rking32) [inline plugins/anilist.py:496]:
```suggestion
        await message.err(f"[{error_sts}]")
```

File: plugins/anilist.py [hunk 18]
```diff
@@ -0,0 +1,638 @@
...
+        progress_args=(
+            "Downloading Media", userge, message, c_time
+        )
+    )
+    dls_loc = os.path.join(Config.DOWN_PATH, os.path.basename(dls))
+    if replied.animation or replied.video:
+        img_loc = os.path.join(Config.DOWN_PATH, "trace.png")
+        await take_screen_shot(dls_loc, 0, img_loc)
+        if not os.path.lexists(img_loc):
+            await message.err("`Media not found...`", del_in=5)
```
Comment (rking32) [inline plugins/anilist.py:581]:
```suggestion
            await message.err("Media not found...", del_in=5)
```

**Comment by Phyco-Ninja** [comment]
Kek. But how to disable that bug risk warns

**Pull request merged**